File: Kotlin/Shortest_Unsorted_Continuous_Subarray.kt
class Solution {
    fun findUnsortedSubarray(nums: IntArray): Int {
        val n = nums.size
        var start = 0
        var end = n - 1
        
        while (start < n - 1 && nums[start] <= nums[start + 1]) {
            start++
        }
        if (start == n - 1) return 0
        
        while (end > 0 && nums[end] >= nums[end - 1]) {
            end--
        }
        
        val subMin = nums.slice(start..end).minOrNull()!!
        val subMax = nums.slice(start..end).maxOrNull()!!
        
        while (start > 0 && nums[start - 1] > subMin) {
            start--
        }
        while (end < n - 1 && nums[end + 1] < subMax) {
            end++
        }
        
        return end - start + 1
    }
}

File: Golang/Shortest_Unsorted_Continuous_Subarray.go
func findUnsortedSubarray(nums []int) int {
	n := len(nums)
	maxVal, minVal := nums[0], nums[n-1]
	start, end := -1, -2
	for i := 1; i < n; i++ {
		maxVal = max(maxVal, nums[i])
		minVal = min(minVal, nums[n-1-i])
		if nums[i] < maxVal {
			end = i
		}
		if nums[n-1-i] > minVal {
			start = n - 1 - i
		}
	}
	return end - start + 1
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}